File: 2025-09-September-GFG-POTD/September 20 - Longest Subarray Length/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    static int longestSubarray(vector<int>& arr) {
        int n = arr.size();
        vector<int> nextGreater(n, n);
        vector<int> prevGreater(n, -1);
        stack<int> st;

        // Next Greater Element to the Right
        for (int i = 0; i < n; i++) {
            while (!st.empty() && arr[st.top()] < arr[i]) {
                nextGreater[st.top()] = i;
                st.pop();
            }
            st.push(i);
        }

        // Clear stack for prev greater
        while (!st.empty()) st.pop();

        // Next Greater Element to the Left
        for (int i = n - 1; i >= 0; i--) {
            while (!st.empty() && arr[st.top()] < arr[i]) {
                prevGreater[st.top()] = i;
                st.pop();
            }
            st.push(i);
        }

        int maxLength = 0;
        for (int i = 0; i < n; i++) {
            int windowSize = nextGreater[i] - prevGreater[i] - 1;
            if (windowSize >= arr[i]) {
                maxLength = max(maxLength, windowSize);
            }
        }
        return maxLength;
    }
};

// Example usage
int main() {
    vector<int> arr = {2, 2, 3, 15, 8, 11, 6, 6};
    cout << Solution::longestSubarray(arr) << endl; // Expected 3
    return 0;
}

File: 2025-09-September-GFG-POTD/September 20 - Longest Subarray Length/Solution.py
from typing import List

class Solution:
    @staticmethod
    def longestSubarray(arr: List[int]) -> int:
        n = len(arr)
        next_greater = [n] * n
        prev_greater = [-1] * n
        st = []

        # Next Greater Element to the Right
        for i in range(n):
            while st and arr[st[-1]] < arr[i]:
                next_greater[st.pop()] = i
            st.append(i)

        st.clear()

        # Next Greater Element to the Left
        for i in range(n - 1, -1, -1):
            while st and arr[st[-1]] < arr[i]:
                prev_greater[st.pop()] = i
            st.append(i)

        max_length = 0
        for i in range(n):
            window_size = next_greater[i] - prev_greater[i] - 1
            if window_size >= arr[i]:
                max_length = max(max_length, window_size)

        return max_length


# Example usage
arr = [2, 2, 3, 15, 8, 11, 6, 6]
print(Solution.longestSubarray(arr))  # Expected 3
